File: iOS/Tyche/Tyche/UI/Profile/ProfileViewModel.swift
import Account
import Combine
import Foundation
import Session
import UIKit
import ViewingState

enum ProfileAvatarSource: Equatable {
    case local(UIImage)
    case remote(URL)
    case letter
}

@MainActor
class ProfileViewModel: ObservableObject {
    @Published private(set) var accountId: String = ""
    @Published private(set) var username: String = ""
    @Published private(set) var email: String = ""
    @Published private(set) var uploadState: LoadState<Void> = .idle

    private let currentAccountModel: CurrentAccountModel
    private let currentAccountCoordinator: CurrentAccountCoordinator
    private let onUploadAvatar: (Data) async -> Result<Void, Error>

    private var pendingUpload: (image: UIImage, data: Data)?
    private var cancellables = Set<AnyCancellable>()

    init(
        currentAccountModel: CurrentAccountModel,
        currentAccountCoordinator: CurrentAccountCoordinator,
        onUploadAvatar: @escaping (Data) async -> Result<Void, Error>
    ) {
        self.currentAccountModel = currentAccountModel
        self.currentAccountCoordinator = currentAccountCoordinator
        self.onUploadAvatar = onUploadAvatar

        currentAccountModel.$account
            .sink { [weak self] account in
                self?.accountId = account?.accountId ?? ""
                self?.username = account?.username ?? ""
                self?.email = account?.email ?? ""
            }
            .store(in: &cancellables)
    }

    /// The uploading photo shows optimistically; a failed upload falls back to what was
    /// there before, and an account without a photo falls back to the letter avatar.
    var avatarSource: ProfileAvatarSource {
        if uploadState.isLoading(), let pendingUpload {
            return .local(pendingUpload.image)
        }

        if !accountId.isEmpty, let url = AvatarURL.of(accountId: accountId) {
            return .remote(url)
        }

        return .letter
    }

    func refreshAccount() async {
        await currentAccountCoordinator.refresh(trigger: .profileOpened)
    }

    @discardableResult
    func beginUpload(image: UIImage, data: Data) -> Task<Void, Never> {
        pendingUpload = (image, data)
        return performUpload()
    }

    @discardableResult
    func retryUpload() -> Task<Void, Never> {
        performUpload()
    }

    func dismissUploadError() {
        pendingUpload = nil
        uploadState = .idle
    }

    private func performUpload() -> Task<Void, Never> {
        uploadState = .loading

        return Task {
            guard let pending = pendingUpload else {
                uploadState = .idle
                return
            }

            let result = await onUploadAvatar(pending.data)

            switch result {
            case .success:
                // The upload flow has already seeded the shared avatar store, so every
                // visible surface shows the new photo before this state lands.
                pendingUpload = nil
                uploadState = .loaded(())
            case .failure(let error):
                uploadState = .failure(error)
            }
        }
    }
}

File: Android/app/src/main/java/com/felipearpa/tyche/profile/ProfileViewModel.kt
package com.felipearpa.tyche.profile

import android.graphics.Bitmap
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.felipearpa.tyche.account.AvatarUrl
import com.felipearpa.tyche.session.CurrentAccountCoordinator
import com.felipearpa.tyche.session.CurrentAccountRefreshTrigger
import com.felipearpa.tyche.session.avatar.application.UploadAvatar
import com.felipearpa.ui.state.LoadState
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

sealed class ProfileAvatarSource {
    data class Local(val bitmap: Bitmap) : ProfileAvatarSource()
    data class Remote(val url: String) : ProfileAvatarSource()
    data object Letter : ProfileAvatarSource()
}

class ProfileViewModel(
    private val uploadAvatar: UploadAvatar,
    private val currentAccountCoordinator: CurrentAccountCoordinator,
) : ViewModel() {
    val accountId: StateFlow<String> = currentAccountCoordinator.state
        .map { bundle -> bundle?.accountId.orEmpty() }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.Eagerly,
            initialValue = currentAccountCoordinator.state.value?.accountId.orEmpty(),
        )

    val username: StateFlow<String> = currentAccountCoordinator.state
        .map { bundle -> bundle?.username.orEmpty() }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.Eagerly,
            initialValue = currentAccountCoordinator.state.value?.username.orEmpty(),
        )

    val email: StateFlow<String> = currentAccountCoordinator.state
        .map { bundle -> bundle?.email.orEmpty() }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.Eagerly,
            initialValue = currentAccountCoordinator.state.value?.email.orEmpty(),
        )

    private val _uploadState = MutableStateFlow<LoadState<Unit>>(LoadState.Idle)
    val uploadState: StateFlow<LoadState<Unit>> = _uploadState.asStateFlow()

    private var pendingUpload: Pair<Bitmap, ByteArray>? = null

    /**
     * The uploading photo shows optimistically; a failed upload falls back to what was there
     * before, and an account without a photo falls back to the letter avatar. After a
     * successful upload the remote branch is already served from the seeded shared store, so
     * no view-local copy is retained.
     */
    val avatarSource: StateFlow<ProfileAvatarSource> = combine(
        currentAccountCoordinator.state,
        _uploadState,
    ) { bundle, uploadState ->
        val pending = pendingUpload

        when {
            uploadState is LoadState.Loading && pending != null ->
                ProfileAvatarSource.Local(pending.first)

            !bundle?.accountId.isNullOrEmpty() ->
                ProfileAvatarSource.Remote(AvatarUrl.of(bundle!!.accountId))

            else -> ProfileAvatarSource.Letter
        }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.Eagerly,
        initialValue = ProfileAvatarSource.Letter,
    )

    init {
        viewModelScope.launch {
            currentAccountCoordinator.refresh(CurrentAccountRefreshTrigger.PROFILE_OPENED)
        }
    }

    fun beginUpload(bitmap: Bitmap, imageData: ByteArray): Job {
        pendingUpload = bitmap to imageData
        return performUpload()
    }

    fun retryUpload(): Job = performUpload()

    fun dismissUploadError() {
        pendingUpload = null
        _uploadState.value = LoadState.Idle
    }

    private fun performUpload(): Job {
        _uploadState.value = LoadState.Loading

        return viewModelScope.launch {
            val pending = pendingUpload ?: run {
                _uploadState.value = LoadState.Idle
                return@launch
            }

            uploadAvatar.execute(imageData = pending.second)
                .onSuccess {
                    // The upload flow already seeded the shared avatar store, so success only
                    // clears the pending photo and reports loaded.
                    pendingUpload = null
                    _uploadState.value = LoadState.Loaded(Unit)
                }
                .onFailure { exception ->
                    _uploadState.value = LoadState.Failure(exception)
                }
        }
    }
}
